### ISSUE TITLE
rxdart 0.27.2 introduces a breaking change to BehaviorSubject.map

### ISSUE THREAD

**Issue by ryanheise** [ReactiveX/rxdart#625]
In rxdart 0.27.1, `BehaviorSubject.map` returns a `ValueStream` but this breaks in 0.27.2 which returns a `Stream` (a minor version increment). I like being able to map a `ValueStream` into another `ValueStream`, so maybe this interface should be defined as part of the `ValueStream` API, but in any case I think the behaviour should at least be brought back to `BehaviorSubject` to avoid the breaking change.

**Comment by hoc081098** [comment]
- That is just a workaround to keep behavior of `BehaviorSubject` (replay the latest value when listening) when using built-in operators.
- And that is an mistake to expose return type as `ValueStream`
- And returned `ValueStream`s have not been intented to use directly, it causes many problems
```dart
Before:
var s = BehaviorSubjec.seedee(1);
print(s.value) // 1
ValueStream<int> mapped = s.map((v) => v + 1);
print(mapped.value); // throws Error
```

**Comment by ryanheise** [comment]
That said, it is a breaking change with no prior indication that this was an internal method, so it breaks apps/plugins that were previously depending on it. I think the correct way to remove this method would be to first deprecate it and remove it in the next major version.

But I think a better solution is to improve the semantics of this method rather than remove it in the next release. i.e.

* If the current value stream has a value, return `shareValueSeeded`
* else return `shareValue`

**Comment by hoc081098** [comment]
That is method overriding, cannot deprecated them.

**Comment by ryanheise** [comment]
Ah, I didn't think of that. _Although_ the solution I really prefer doesn't require deprecation anyway ;-) i.e. Keep the override so as to avoid any breaking change on a minor release, and then pave the pathway to support a specialised `map` for all `ValueStream`s with better semantics.

**Comment by alexgarbarev** [comment]
Yeah, that change broke my app as well. I wasn't aware of all methods available in rxdart, and I read "changelog" after upgrade, but I couldn't figure out how to fix it, before I found this thread.

I believe that for this change, we should add a note in a "changelog", and suggest a workaround with correct syntax as mentioned in there https://github.com/ReactiveX/rxdart/issues/625#issuecomment-916518794

btw thanks for the library overall :)

**Comment by hoc081098** [comment]
@alexgarbarev 
```dart
final s = BehaviorSubject.seeded(1);

Future<String> converter(int v) {  }

ValueStream<String> mapped = s.asyncExpand(converter)
.cast<String?>()
.shareValueSeeded(null); // null means loading state

String? value = mapped.valueOrNull; 
if (value == null) {} // loading
else {}

mapped.listen((String? value) {
   if (value == null) {} // loading
   else {}
});

```
